### ISSUE TITLE
Error : TypeError: Cannot read property 'isEmpty' of undefined

### ISSUE THREAD

**Issue by abarre** [css/csso#369]
I have a file containing a CSS portion that contains an error : 

```css
.k-carousel-slide img{min-width:0px;display:block;}}.container{width:1258px;margin:0 auto;}
```
You cann see the two `}` that crashes csso.

In this case, csso will completely fails even if it uses csstree (that is tolerant to this error).

Stack trace : 
```
TypeError: Cannot read property 'isEmpty' of undefined
    at Object.cleanRuleset (csso/lib/clean/Rule.js:85:30)
    at Object.walk.leave (csso/lib/clean/index.js:16:37)
    at walkNode (csso/node_modules/css-tree/lib/walker/create.js:161:19)
    at List.each (csso/node_modules/css-tree/lib/utils/list.js:159:12)
    at Object.eval [as StyleSheet] (eval at createTypeIterator (csso/node_modules/css-tree/lib/walker/create.js:112:12), <anonymous>:5:15)
    at walkNode (csso/node_modules/css-tree/lib/walker/create.js:158:37)
    at walk (csso/node_modules/css-tree/lib/walker/create.js:214:9)
    at module.exports (csso/lib/clean/index.js:13:5)
    at compressChunk (csso/lib/compress.js:69:5)
    at compress (csso/lib/compress.js:154:9)
```
I think that the issue is related to the fact that the type `Raw` is not handled by csso.

Node that crashes csso : 
```js
{ type: 'Rule',
  loc: null,
  prelude: { type: 'Raw', loc: null, value: '}.container' },
  block:
   { type: 'Block',
     loc: null,
     children:
      { cursor: null,
        head:
         { prev: null,
           next:
            { prev: [Circular],
              next: null,
              data:
               { type: 'Declaration',
                 loc: null,
                 important: false,
                 property: 'margin',
                 value:
                  { type: 'Value',
                    loc: null,
                    children:
                     { cursor: null,
                       head:
                        { prev: null,
                          next:
                           { prev: [Circular],
                             next:
                              { prev: [Circular],
                                next: null,
                                data: { type: 'Identifier', loc: null, name: 'auto' } },
                             data: { type: 'WhiteSpace', loc: null, value: ' ' } },
                          data: { type: 'Number', loc: null, value: '0' } },
                       tail:
                        { prev:
                           { prev:
                              { prev: null,
                                next: [Circular],
                                data: { type: 'Number', loc: null, value: '0' } },
                             next: [Circular],
                             data: { type: 'WhiteSpace', loc: null, value: ' ' } },
                          next: null,
                          data: { type: 'Identifier', loc: null, name: 'auto' } } } } } },
           data:
            { type: 'Declaration',
              loc: null,
              important: false,
              property: 'width',
              value:
               { type: 'Value',
                 loc: null,
                 children:
                  { cursor: null,
                    head:
                     { prev: null,
                       next: null,
                       data: { type: 'Dimension', loc: null, value: '1258', unit: 'px' } },
                    tail:
                     { prev: null,
                       next: null,
                       data: { type: 'Dimension', loc: null, value: '1258', unit: 'px' } } } } } },
        tail:
         { prev:
            { prev: null,
              next: [Circular],
              data:
               { type: 'Declaration',
                 loc: null,
                 important: false,
                 property: 'width',
                 value:
                  { type: 'Value',
                    loc: null,
                    children:
                     { cursor: null,
                       head:
                        { prev: null,
                          next: null,
                          data: { type: 'Dimension', loc: null, value: '1258', unit: 'px' } },
                       tail:
                        { prev: null,
                          next: null,
                          data: { type: 'Dimension', loc: null, value: '1258', unit: 'px' } } } } } },
           next: null,
           data:
            { type: 'Declaration',
              loc: null,
              important: false,
              property: 'margin',
              value:
               { type: 'Value',
                 loc: null,
                 children:
                  { cursor: null,
                    head:
                     { prev: null,
                       next:
                        { prev: [Circular],
                          next:
                           { prev: [Circular],
                             next: null,
                             data: { type: 'Identifier', loc: null, name: 'auto' } },
                          data: { type: 'WhiteSpace', loc: null, value: ' ' } },
                       data: { type: 'Number', loc: null, value: '0' } },
                    tail:
                     { prev:
                        { prev:
                           { prev: null,
                             next: [Circular],
                             data: { type: 'Number', loc: null, value: '0' } },
                          next: [Circular],
                          data: { type: 'WhiteSpace', loc: null, value: ' ' } },
                       next: null,
                       data: { type: 'Identifier', loc: null, name: 'auto' } } } } } } } } }
```

**Comment by horgh** [comment]
I have a similar crash when there is a trailing comma before `{`:

```css
.content-wrap div.paragraph ul, { padding-left: 2em !important; }
.content-wrap div.paragraph li, { padding-left: 0 !important; margin: 0 !important; }
```

```
$ csso -v
3.5.0
$ csso style.css > min.css
/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/index.js:289
            throw e;
            ^

TypeError: Cannot read property 'isEmpty' of undefined
    at Object.cleanRuleset (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/csso/lib/clean/Rule.js:83:31)
    at Object.leave (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/csso/lib/clean/index.js:16:37)
    at List.walkNode (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/css-tree/lib/walker/create.js:161:19)
    at List.each (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/css-tree/lib/utils/list.js:159:12)
    at Object.eval [as StyleSheet] (eval at createTypeIterator (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/css-tree/lib/walker/create.js:112:12), <anonymous>:5:15)
    at walkNode (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/css-tree/lib/walker/create.js:158:37)
    at walk (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/css-tree/lib/walker/create.js:214:9)
    at module.exports (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/csso/lib/clean/index.js:13:5)
    at compressChunk (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/csso/lib/compress.js:68:5)
    at compress (/home/will/bin/node-v8.11.1-linux-x64/lib/node_modules/csso-cli/node_modules/csso/lib/compress.js:153:9)
```

**Comment by hansott** [comment]
Another crash:

```css
@media screen and (min-width: 0 \0) {
  .body {
    color: black;
  }
}
```

Version: 3.5.1

```
/project/node_modules/csso-cli/index.js:289
            throw e;
            ^

TypeError: Cannot read property 'isEmpty' of undefined
    at Object.cleanAtrule (/project/node_modules/csso/lib/clean/Atrule.js:61:60)
    at Object.leave (/project/node_modules/csso/lib/clean/index.js:16:37)
    at List.walkNode (/project/node_modules/css-tree/lib/walker/create.js:161:19)
    at List.each (/project/node_modules/css-tree/lib/utils/list.js:159:12)
    at Object.eval [as StyleSheet] (eval at createTypeIterator (/project/node_modules/css-tree/lib/walker/create.js:112:12), <anonymous>:5:15)
    at walkNode (/project/node_modules/css-tree/lib/walker/create.js:158:37)
    at walk (/project/node_modules/css-tree/lib/walker/create.js:214:9)
    at module.exports (/project/node_modules/csso/lib/clean/index.js:13:5)
    at compressChunk (/project/node_modules/csso/lib/compress.js:68:5)
    at compress (/project/node_modules/csso/lib/compress.js:153:9)
```

**Comment by jt3k** [comment]
same error ((

**Comment by Tom910** [comment]
Some problem. Now it is very difficult to find in what code the error

**Comment by MaximeScibetta** [comment]
same issue

**Comment by tavishaggarwal** [comment]
Any fix on this?

**Comment by itsHall** [comment]
I was having the same issue - my fix was to change the gulp-csso version from v^3.0.0 to v2.0.0. (I am using gulp v3.9.1).

**Comment by RyanPWalker** [comment]
Any update on this?  This is still a current bug.

**Comment by lahmatiy** [comment]
@abarre Fixed in 4.0.2. Notice that `.container{ ... }` block is removing since unpaired `}` breaks selector
@horgh Your case is also fixed. Notice that both rules will be removed since selectors are invalid.
@hansott CSSO is not crash in your case. However, at-rule will be removed, because `(min-width: 0 \0)` is invalid. To fix that change it to `(min-width: 0\0)` (remove a whitespace before `\0`) – in this case hack will still work and at-rule will not be removed.
